src/bitwarden_workflow_linter/rules/underscore_outputs.py

```diff
@@ -0,0 +1,110 @@
+            required anywhere in the application.
+        """
+        self.message = "outputs with more than one word must use an underscore"
+        self.on_fail = LintLevels.WARNING
+        self.compatibility = [Workflow, Job, Step]
+        self.settings = settings
+
+    def fn(self, obj: Union[Workflow, Job, Step]) -> Tuple[bool, str]:
+        """Enforces all outputs to have an underscore in the key name.
+
+        This Rule checks all outputs in a Workflow, Job, or Step to ensure that
+        the key name contains an underscore. This is to ensure that the naming
+        convention is consistent across all outputs in the workflow configuration
+
+        Example:
+        ---
+        on:
+          workflow_dispatch:
+            outputs:
+              registry:
+                value: 'Test Value'
+              some_registry:
+                value: 'Test Value'
+          workflow_call:
+            outputs:
+              registry:
+                value: 'Test Value'
+              some_registry:
+                value: 'Test Value'
+        jobs:
+          job-key:
+            runs-on: ubuntu-22.04
+            outputs:
+              test_key_job: ${{ steps.test_output_1.outputs.test_key }}
+            steps:
+              - name: Test output in one-line run step
+                id: test_output_1
+                run: echo "test_key_1=Test-Value1" >> $GITHUB_OUTPUT
+
+              - name: Test output in multi-line run step
+                id: test_output_2
+                run: |
+                  echo
+                  fake-command=Test-Value2
+                  echo "test_key_2=$REF" >> $GITHUB_OUTPUT
+                  echo "deployed_ref=$DEPLOYED_REF" >> $GITHUB_OUTPUT
+
+              - name: Test step with no run
+                id: test_output_3
+                uses: actions/checkout@v2
+                with:
+                  ref: ${{ github.ref }}
+                  fetch-depth: 0
+
+        In this example, in workflow level 'registry' and 'some_registry' are outputs
+        that satisfy the rule in both 'workflow_dispatch' and 'workflow_call' events.
+        In job level 'test_key_job' satisfies the rule.
+        In step level 'test_key_1', 'test_key_2', and 'deployed_ref' satisfy the rule.
+
+        See tests/rules/test_underscore_outputs.py for incorrect examples.
+        """
+
+        outputs = []
+
+        if isinstance(obj, Workflow):
+            if obj.on.get("workflow_dispatch"):
+                outputs.extend(obj.on["workflow_dispatch"]["outputs"].keys())
+
+            if obj.on.get("workflow_call"):
+                outputs.extend(obj.on["workflow_call"]["outputs"].keys())
+
+        if isinstance(obj, Job):
+            if obj.outputs:
+                outputs.extend(obj.outputs.keys())
+
+        if isinstance(obj, Step):
+            if obj.run:
+                outputs.extend(re.findall(r"\b([a-zA-Z0-9_-]+)\s*=\s*[^=]*>>", obj.run))
```

Comment:
Can we add some tests around `steps.[*].run` using steps like the following? Is the output by the new rule what we expect?


```yaml
- name: Test Step
  run: |
    echo "this is a test value: does-it-break?" >> /tmp/test
    
- name: Test Step 2
  run: |
    TEST_FILE=/tmp/test2
    echo "this is a test value: does-it-break?" >> $TEST_FILE
```


Sorry, I thought I caught this in the first review.

Reply:
What are we testing here please?

Reply:
The test cases here are testing the regex logic to validate that the rule only warns on step output specifically rather than a legitimate echo statement that the workflow engineer needs.

Reply:
Oh. well, the regex is specifically crafted to match step outputs in the format we usually write. It can't match the above tests because of the presence of `:` and `whitespaces` inbetween the variable sides. I can however improve the regex to include `$GITHUB_OUTPUT`. That should make our match incorruptible within the scripts